x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/categorization/CategorizeTextAggregator.java

```diff
@@ -0,0 +1,245 @@
+            results[ordIdx] = new InternalCategorizationAggregation(
+                name,
+                bucketCountThresholds.getRequiredSize(),
+                bucketCountThresholds.getMinDocCount(),
+                maxUniqueTokens,
+                maxMatchTokens,
+                similarityThreshold,
+                metadata(),
+                Arrays.asList(bucketArray)
+            );
+        }
+        return results;
+    }
+
+    @Override
+    public InternalAggregation buildEmptyAggregation() {
+        return new InternalCategorizationAggregation(
+            name,
+            bucketCountThresholds.getRequiredSize(),
+            bucketCountThresholds.getMinDocCount(),
+            maxUniqueTokens,
+            maxMatchTokens,
+            similarityThreshold,
+            metadata()
+        );
+    }
+
+    @Override
+    protected LeafBucketCollector getLeafCollector(LeafReaderContext ctx, LeafBucketCollector sub) throws IOException {
+        return new LeafBucketCollectorBase(sub, null) {
+            private final BytesRefBuilder scratch = new BytesRefBuilder();
+
+            @Override
+            public void collect(int doc, long owningBucketOrd) throws IOException {
+                collectFromSource(doc, owningBucketOrd);
+            }
+
+            private void collectFromSource(int doc, long owningBucketOrd) throws IOException {
+                sourceLookup.setSegmentAndDocument(ctx, doc);
+                Iterator<String> itr = sourceLookup.extractRawValues(sourceFieldName).stream().map(obj -> {
+                    if (obj == null) {
+                        return null;
+                    }
+                    if (obj instanceof BytesRef) {
+                        return fieldType.valueForDisplay(obj).toString();
+                    }
+                    return obj.toString();
+                }).iterator();
+                while (itr.hasNext()) {
+                    TokenStream ts = analyzer.tokenStream(fieldType.name(), itr.next());
+                    processTokenStream(owningBucketOrd, ts, doc);
+                }
+            }
+
+            private void processTokenStream(long owningBucketOrd, TokenStream ts, int doc) throws IOException {
+                ArrayList<Integer> tokens = new ArrayList<>();
+                try {
+                    CharTermAttribute termAtt = ts.addAttribute(CharTermAttribute.class);
+                    ts.reset();
+                    while (ts.incrementToken()) {
+                        tokens.add(bytesRefHash.put(new BytesRef(termAtt)));
+                    }
+                    if (tokens.isEmpty()) {
+                        return;
+                    }
+                } finally {
+                    ts.close();
+                }
+                categorizers = bigArrays().grow(categorizers, owningBucketOrd + 1);
+                CategorizationTokenTree categorizer = categorizers.get(owningBucketOrd);
```

Comment:
Could you look up the `CategorizationTokenTree` before calling this method? Just once per value in the source? It won't amount to any speed change, but I think it's a little more obvious to read.